zap/src/output/server.rs

```diff
@@ -1,434 +1,514 @@
-				.ev_decls
-				.iter()
-				.enumerate()
-				.filter(|(_, ev_decl)| ev_decl.from == EvSource::Client && ev_decl.evty == EvType::Unreliable)
-			{
-				let id = i + 1;
-
-				if i == 0 {
-					writeln!(f, "\tif id == {id} then")?;
-				} else {
-					writeln!(f, "\telseif id == {id} then")?;
-				}
-
-				writeln!(f, "\t\tlocal value;")?;
-
-				for stmt in gen_des(&ev_decl.data, "value", true) {
-					writeln!(f, "\t\t{stmt}")?;
-				}
-
-				match ev_decl.call {
-					EvCall::SingleSync => writeln!(f, "\t\tif events[{id}] then events[{id}](player, value) end")?,
-					EvCall::SingleAsync => writeln!(
-						f,
-						"\t\tif events[{id}] then task.spawn(events[{id}], player, value) end"
-					)?,
-
-					EvCall::ManySync => writeln!(f, "\t\tfor _, cb in events[{id}] do cb(player, value) end")?,
-					EvCall::ManyAsync => {
-						writeln!(f, "\t\tfor _, cb in events[{id}] do task.spawn(cb, player, value) end")?
-					}
-				}
-
-				writeln!(f, "\t\tend")?;
-			}
+			let id = i + 1;
 
-			writeln!(f, "\tend")?;
-			writeln!(f, "end)")?;
+			self.push_unreliable_callback(first, ev, id);
+			first = false;
 		}
 
-		// Output Event Declarations
-		for (i, _) in file.ev_decls.iter().enumerate().filter(|(_, ev_decl)| {
+		self.push_unreliable_footer();
+	}
+
+	fn push_callback_lists(&mut self) {
+		for (i, _) in self.file.ev_decls.iter().enumerate().filter(|(_, ev_decl)| {
 			ev_decl.from == EvSource::Client && matches!(ev_decl.call, EvCall::ManyAsync | EvCall::ManySync)
 		}) {
 			let id = i + 1;
 
-			writeln!(f, "events[{id}] = {{}}")?;
+			self.push_line(&format!("events[{id}] = {{}}"));
 		}
+	}
+
+	fn push_return_fire(&mut self, ev: &EvDecl, id: usize) {
+		let ty = &ev.data;
+
+		let fire = casing(self.file.casing, "Fire", "fire", "fire");
+		let player = casing(self.file.casing, "Player", "player", "player");
+		let value = casing(self.file.casing, "Value", "value", "value");
+
+		self.push_line(&format!("{fire} = function({player}: Player, {value}: {ty})"));
+		self.indent();
 
-		writeln!(f, "return {{")?;
+		match ev.evty {
+			EvType::Reliable => self.push_line(&format!("load(player_map[{player}])")),
+			EvType::Unreliable => self.push_line("load_empty()"),
+		}
+
+		self.push_line(&format!("buffer.write{}(outgoing_buff, {id})", self.file.event_id_ty()));
+		self.push_stmts(&gen_ser(ty, value.into(), self.file.write_checks));
 
-		// Output Fire Return
-		let value = casing(file.casing, "Value", "value", "value");
+		match ev.evty {
+			EvType::Reliable => self.push_line(&format!("player_map[{player}] = save()")),
+			EvType::Unreliable => {
+				self.push_line("local buff = buffer.create(outgoing_used)");
+				self.push_line("buffer.copy(buff, 0, outgoing_buff, 0, outgoing_used)");
+				self.push_line(&format!("unreliable:FireClient({player}, buff, outgoing_inst)"));
+			}
+		}
 
-		for (i, ev_decl) in file
+		self.dedent();
+		self.push_line("end,");
+	}
+
+	fn push_return_fire_all(&mut self, ev: &EvDecl, id: usize) {
```

Comment:
I wonder if `push_return_fire_*` should inherit from a base fuction to save some duplication

Reply:
I don't know if this is possible in rust, if it is I'll happily look into it.

Reply:
Should've been clearer, by inheriting I mean just calling a base one in each of these